### PR TITLE
C3 sign in out

### PR DESCRIPTION
**Pull request by loicteixeira**
Not much to do for this one as devise (which was added during the previous checkpoint) already had the session implemented. It was just about testing then (and testing the password reset was quite interesting).

#### User Stories
- As a user, I want to sign in and out of Blocitoff

#### Required Tests
- Sign into Blocitoff, does the top navigation change to indicate you are signed in?
- Sign out of Blocitoff, does the top navigation change to indicate you are signed out?
- What happens if you attempt to reset your password?

### CODE REVIEW

File: app/views/layouts/application.html.erb [hunk 1]
```diff
@@ -11,6 +11,15 @@
 <p class="notice"><%= notice %></p>
 <p class="alert"><%= alert %></p>
 
+<div class="navigation">
+  <% if current_user %>
```
Comment (eliotsykes) [inline app/views/layouts/application.html.erb]: Minor thing to think about here - devise provides the `current_user` method, and also provides a `user_signed_in?` method which could do the same job as `current_user` in this case. The benefit of `if user_signed_in?` is its language more clearly reveals the intent of the `if` block, compared to `if current_user`:

```html
<% if current_user %>
  The intent here is to show something to users who are signed in
<% else %>
...
```
compared to
```html
<% if user_signed_in? %>
  The intent here is to show something to users who are signed in
<% else %>
...
```

File: spec/features/user_sign_in_out_spec.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
+require 'rails_helper'
+
+feature 'when user is signed out' do
+  let(:user) { create(:user) }
+
+  scenario 'the navigation shows a sign in link' do
+    visit root_path
+    find_link('Sign In')
```
Comment (eliotsykes) [inline spec/features/user_sign_in_out_spec.rb]: I'm not sure line 8 will fail if the link isn't present, favor using `expect`:
```ruby
expect(page).to have_link 'Sign In'
```
Consider adding an expectation that the Sign Out link is not shown.

File: spec/features/user_sign_in_out_spec.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  scenario 'is able to reset password' do
+    visit new_user_session_path
+    click_link 'Forgot your password?'
+
+    fill_in 'Email', with: user.email
+    click_button 'Send me reset password instructions'
+
+    last_email = ActionMailer::Base.deliveries.last
```
Comment (eliotsykes) [inline spec/features/user_sign_in_out_spec.rb]: Awesome work figuring out how to test emails in feature specs, this is an absolutely fine way of doing these tests.

You might be interested in the [email_spec gem](https://github.com/bmabey/email-spec). It provides some well-named methods like `open_email` and `visit_in_email` to help interact with emails in a straightforward, readable way. See these resources for more info on setup and usage:

- https://github.com/eliotsykes/rspec-rails-examples/blob/master/spec/support/email_spec.rb
- https://github.com/eliotsykes/rspec-rails-examples#email-spec-examples
- https://github.com/eliotsykes/rspec-rails-examples/blob/master/spec/features/user_registers_spec.rb

File: spec/features/user_sign_in_out_spec.rb [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+  background do
+    visit new_user_session_path
+    fill_in 'Email', with: user.email
+    fill_in 'Password', with: user.password
+    click_button 'Log in'
+  end
+
+  scenario 'the navigation shows a sign out link' do
+    expect(page).to have_content 'Sign Out'
```
Comment (eliotsykes) [inline spec/features/user_sign_in_out_spec.rb]: Consider using the `have_link` matcher for line 44, and adding expectations to this scenario:

- Sign In link is not shown
- Sign in successful message is shown

File: app/views/layouts/application.html.erb (hunk 1 above)
Comment (loicteixeira) [inline app/views/layouts/application.html.erb]: Updated in b1f40feb8300e177e66adc5b5e242063e33542a2.

File: spec/features/user_sign_in_out_spec.rb (hunk 2 above)
Comment (loicteixeira) [inline spec/features/user_sign_in_out_spec.rb]: According to the [documentation](http://www.rubydoc.info/github/jnicklas/capybara/master/Capybara%2FNode%2FFinders%3Afind_link) `find_link` uses `find` which will [raise an error if the element is not found](http://www.rubydoc.info/github/jnicklas/capybara/master/Capybara%2FNode%2FFinders%3Afind). However I agree that the intend of `expect(page).to have_link '<text>'` is clearer than `find_link('<text>')` and will favour it in the future.

Updated in 815805b515ae16453f8d68e251084a4ffe0b4a2b.

File: spec/features/user_sign_in_out_spec.rb (hunk 4 above)
Comment (loicteixeira) [inline spec/features/user_sign_in_out_spec.rb]: Updated in 815805b515ae16453f8d68e251084a4ffe0b4a2b.

File: spec/features/user_sign_in_out_spec.rb (hunk 3 above)
Comment (loicteixeira) [inline spec/features/user_sign_in_out_spec.rb]: Ahah that would have been so much easier and much more readable! I'll give it a try next time.

**Pull request merged**